File: search_rotated_array.cpp
// Time Complexity : O(log n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no

class Solution {
public:
    int search(vector<int>& nums, int target) {
       int low = 0;
       int high = nums.size()-1;
       //to avoid the overflow we use this method
       

       while(low <= high) {
           int mid = low + (high - low)/2;
           if(nums[mid] == target) return mid;
           //check if left is sorted
           if(nums[low] <= nums[mid]) {
               //check if target is in this range
               if(nums[low] <= target && target <= nums[mid]) {
                   high = mid-1; 
               } else {
                   low = mid+1;
               }
           } else {
               //check if target is in this range
               if(nums[mid] <= target && target <= nums[high]) {
                 low = mid + 1;
                } else {
                  high = mid - 1;  
                } 
           }
          
       }
        return -1;
    }
};

File: SearchRotatedArray.java
// Time Complexity :O(logN)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this : No
public class SearchRotatedArray {
    public int search(int[] nums, int target) {
        int low = 0;
        int high = nums.length-1;
        if(nums==null||nums.length==0)
            return -1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid]==target)
                return mid;
            if(nums[mid]>=nums[low]){
                if(target>=nums[low]&&target<nums[mid]){
                    high = mid - 1;
                }else{
                    low  = mid + 1;
                }
            }
            else{
                if(target<=nums[high]&&target>nums[mid]){
                    low = mid + 1;
                }
                else{
                    high = mid - 1;
                }
            }
        }
        return -1;
    }
}
